### PR TITLE
Feature/deployment graph compiler

### PR DESCRIPTION
**Pull request by PawelPeczek-Roboflow**
# Description

Please include a summary of the change and which issue is fixed or implemented. Please also include relevant motivation and context (e.g. links, docs, tickets etc.).

List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

YOUR_ANSWER

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

**Review by paulguerrie — APPROVED** [review]
EPIC PR!  Some small comments/questions but no show stoppers.

File: inference/enterprise/deployments/complier/core.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    DeploymentSpecification,
+)
+from inference.enterprise.deployments.errors import InvalidSpecificationVersionError
+from inference.models.utils import ROBOFLOW_MODEL_TYPES
+
+
+def compile_and_execute(
+    deployment_spec: dict,
+    runtime_parameters: Dict[str, Any],
+    api_key: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/enterprise/deployments/complier/core.py:24]: 👍

File: inference/enterprise/deployments/complier/core.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            runtime_parameters=runtime_parameters,
+            api_key=api_key,
+        )
+    )
+
+
+async def compile_and_execute_async(
+    deployment_spec: dict,
+    runtime_parameters: Dict[str, Any],
+    api_key: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/enterprise/deployments/complier/core.py:41]: 👍

File: inference/enterprise/deployments/complier/graph_parser.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    steps to execute.
+    Let's imagine graph:
+              / -> B -> C -> D
+    A -> IF <             \
+              \ -> E -> F -> G -> H
+    where node G requires node C even though IF branched the execution. In other
+    words - the problem emerges if a node of kind STEP has a parent (node from which
+    it can be achieved) of kind STEP and this parent is in a different branch (point out that
+    we allow for a single step to have multiple steps as input, but they must be at the same
+    execution path - for instance if D requires an output from C and D - this is allowed).
```
[the remark this answers is not in the document]
Comment (PawelPeczek-Roboflow) [reply inference/enterprise/deployments/complier/graph_parser.py:240]: 👍

**Review by paulguerrie — APPROVED** [review]

**Pull request merged**